### DISCUSSION TITLE
New Collection microsoft.ad

### DISCUSSION THREAD

**Discussion by jborean93** [ansible-collections/ansible-inclusion#59] [Resolved reviews]
There is a new collection [microsoft.ad](https://github.com/ansible-collections/microsoft.ad) that has just gone live to Galaxy and Automation Hub. It is a supported collection for managing Active Directory in Ansible. While some of the modules are similar to what's in `ansible.windows` and `community.windows` there will end up being a redirect in place for the `ansible.windows` equivalents but `community.windows` will most likely deprecate their existing ones due to too many differences between what is there and what is in `microsoft.ad`.

Ansible Galaxy: https://galaxy.ansible.com/microsoft/ad
GitHub: https://github.com/ansible-collections/microsoft.ad
Issues Tracker: https://github.com/ansible-collections/microsoft.ad/issues
GitHub Handles: @jborean93
Is the collection part of Automation Hub: Yes
We meet [Collection Requirements](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst): Yes

**Comment by Andersson007** [comment]
@jborean93 hi, thanks for the new collection! I'll post this request in Bullhorn today

**Comment by Andersson007** [comment]
@briantist @lowlydba as you folks are familiar with powershell, could you please take a look? Here's the [checklist](https://github.com/ansible-collections/overview/blob/main/collection_checklist.md).

**Comment by lowlydba** [comment]
Review: Passed ✅ 
---

# Ansible Collections Checklist (short version)

_For details about the following points, refer to the [Collection Requirements](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst)._

Every comment should say whether the reviewer expects it to be addressed, or whether it's optional.

**Public availability and communication:**
- [x] published on [Ansible Galaxy](https://galaxy.ansible.com) with version 1.0.0 or later
- [x] has a policy of releasing, versioning and deprecation announced to contributors and users in some way
   - OPTIONAL: Should explicitly mention semantic versioning & any deprecation policies
- [x] have a Code of Conduct (CoC) compatible with the [Ansible Code of Conduct (CoC)](https://docs.ansible.com/ansible/latest/community/code_of_conduct.html)
- [x] have a publicly available issue tracker that does not require a paid level of service to create an account or view issues
- [x] have a public git repository, releases of the collection are tagged in this repository

**Standards and documentation:**
- [x] adheres to [semantic versioning](https://semver.org/)
- [x] follows [licensing rules](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst#licensing)
- [x] follows the [Ansible documentation standards](https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_documenting.html) and the [style guide](https://docs.ansible.com/ansible/devel/dev_guide/style_guide/index.html#style-guide)
  - [x] FIX: "distibution" typo in group module [docs](https://github.com/ansible-collections/microsoft.ad/blob/main/plugins/modules/group.py#L19)
  - [x] FIX: "traditiional" typo in object_info module [docs](https://github.com/ansible-collections/microsoft.ad/blob/main/plugins/modules/object_info.py#L53)
  - [x] FIX: C() formatting for command reference in object_info module [docs](https://github.com/ansible-collections/microsoft.ad/blob/main/plugins/modules/object_info.py#L71)
  - [x] FIX: "pricipals" typo in user module [docs](https://github.com/ansible-collections/microsoft.ad/blob/main/plugins/modules/user.py#L69)
  - [x] FIX: "memberhips" typo in user module [docs](https://github.com/ansible-collections/microsoft.ad/blob/main/plugins/modules/user.py#L104)
  - [x] FIX: "recommened" typo in ad_object [docs](https://github.com/ansible-collections/microsoft.ad/blob/f155955a3615102bda37843b5175bb68d5e1482e/plugins/doc_fragments/ad_object.py#L153)
- [x] follows [development conventions](https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_best_practices.html); as well as these other requirements:
  - [x] modules satisfy the concept of [idempotency](https://docs.ansible.com/ansible/latest/reference_appendices/glossary.html#term-Idempotency>)
  - [x] modules that only gather information are named `<something>_info`
  - [x] modules that return `ansible_facts` are named `<something>_facts` and do not return non-facts
  - [x] other modules must not allow querying information using specific `state` option values, or similar mechanisms (like `state=get` or `state=query`).  These features should be moved to `<something>_info` or `<something>_fact` modules.
  - [x] `check_mode` is supported in all `_info` and `_facts` modules
- [x] supports Python 2.6 or greater and Python 3.5 or greater. If it does not, read the [full guidelines](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst#python-compatibility) to see if you qualify for an exception and document the unsupported [Python versions](https://docs.ansible.com/ansible/latest/dev_guide/developing_python_3.html#ansible-and-python-3) in the collection ``README.md`` and in every module and plugin (or in doc fragments)
   - OPTIONAL: It looks like only [Python 3.10](https://github.com/ansible-collections/microsoft.ad/blob/db3af6fa86f249752af4761019e4c0acc580f31a/.azure-pipelines/azure-pipelines.yml#L54-L56) is tested against in the pipelines - not sure if testing for 2.6+ is required?
- [x] only uses the [allowed plugin types](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst#modules-plugins) in the `plugins/` directory
- [x] has `README.md`
- [x] documentation, examples, and return sections use FQCNs for the `M(..)` [format macros](https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#linking-and-other-format-macros-within-module-documentation) when referring to modules
- [x] modules (or plugins) from ansible-core use `ansible.builtin.` as a FQCN prefix
- [x] documentation and return sections use `version_added:` containing the *collection* version for which an option, module or plugin was added
- [x] FQCNs are used for `extends_documentation_fragment:`, unless you are referring to doc_fragments from ansible-core
- [x] public plugins, roles and playbooks do not use files outside of `meta/`, `plugins/`, `roles/`, and `playbooks/`

**Collection management:**
- [x] `galaxy.yml` having `tags` field set
- [x] collection dependencies must have a lower bound on the version which is at least 1.0.0, and are all part of the `ansible` package
- [x] `meta/runtime.yml` defines the minimal version of Ansible which the collection works with
- [x] has changelog, preferably with `changelogs/changelog.yaml`
- [x] collection repository should not contain any large objects (binaries) comparatively to the current Galaxy tarball size limit of 20 MB. For example, package installers for testing purposes shouldn't be added.
- [x] collection repository should not contain any unnecessary files like, for example, temporary files. Temporary files should be added to `.gitignore`.

**Tests:**

- [x] passed `ansible-test sanity`
   - OPTIONAL: dlevel sanity is failing currently, but all major versions are ✅ 
- [x] if `test/sanity/ignore*.txt` exists, it MUST not contain error codes listed [here](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst#ci-testing)
- [x] has CI tests up and running against each of the "major versions" of `ansible-base`/`ansible-core` that the collection supports
- [x] all CI tests MUST run against every pull request
- [x] all CI tests MUST run regularly (nightly, or at least once per week)
- [x] sanity tests MUST run against a commit that releases the collection; if they don't pass, the collection won't be released

**Reply by jborean93** [reply]
Thanks for the thorough review, I believe I've addressed the documentation problems in https://github.com/ansible-collections/microsoft.ad/pull/20.

For the Python support the majority of the code is run on the remote side with PowerShell. We have tests covering Server 2012, 2012 R2, 2016, 2019, and 2019 which covers PowerShell 3, 4, and 5.1. There are no remote Python modules and the only Python code is in the action plugins. This has a limited range of supported versions (2.14 uses Python 3.9 - 3.11) and the effort to test this alongside the various PowerShell versions is too much for what we need.

**Reply by Andersson007** [reply]
@lowlydba thanks much for the contribution! Much appreciated!
When you're satisfied, please put something if the collection is OK or not for inclusion. Then we can mark it ready for the second review

**Reply by lowlydba** [reply]
All looks good to me now, thanks! OK for inclusion 👍🏻

**Comment by Andersson007** [comment]
# Ansible Collections Checklist (short version)

_For details about the following points, refer to the [Collection Requirements](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst)._

Every comment should say whether the reviewer expects it to be addressed, or whether it's optional.

Note for reviewers: If you don't know how to check any of the points below, please ask maintainers of the collection you're reviewing or a [Steering Committee member](https://docs.ansible.com/ansible/devel/community/steering/community_steering_committee.html#current-steering-committee-members) for clarifications in comments of corresponding inclusion discussion.

**Public availability and communication:**
- [x] published on [Ansible Galaxy](https://galaxy.ansible.com) with version 1.0.0 or later
- [x] has a policy of releasing, versioning and deprecation announced to contributors and users in some way
  **MUST FIX** I see the process but don't see the policy. It can be in README or in an issue. See community.general as an example.
- [x] have a Code of Conduct (CoC) compatible with the [Ansible Code of Conduct (CoC)](https://docs.ansible.com/ansible/latest/community/code_of_conduct.html)
- [x] have a publicly available issue tracker that does not require a paid level of service to create an account or view issues
- [x] have a public git repository, releases of the collection are tagged in this repository

**Standards and documentation:**
- [x] adheres to [semantic versioning](https://semver.org/)
  **MUST FIX** Looks like it follows but there's no info about it in README
- [x] follows [licensing rules](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst#licensing)
- [ ] follows the [Ansible documentation standards](https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_documenting.html) and the [style guide](https://docs.ansible.com/ansible/devel/dev_guide/style_guide/index.html#style-guide)
  **MUST FIX** please [reflect](https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#documentation-block) in the DOCUMENTATION that the check mode is supported by modules (where supported)
- [x] follows [development conventions](https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_best_practices.html); as well as these other requirements:
  - [x] modules satisfy the concept of [idempotency](https://docs.ansible.com/ansible/latest/reference_appendices/glossary.html#term-Idempotency>)
  - [x] modules that only gather information are named `<something>_info`
  - [x] modules that return `ansible_facts` are named `<something>_facts` and do not return non-facts
  - [x] other modules must not allow querying information using specific `state` option values, or similar mechanisms (like `state=get` or `state=query`).  These features should be moved to `<something>_info` or `<something>_fact` modules.
  - [x] `check_mode` is supported in all `_info` and `_facts` modules
- [x] supports Python 2.6 or greater and Python 3.5 or greater. If it does not, read the [full guidelines](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst#python-compatibility) to see if you qualify for an exception and document the unsupported [Python versions](https://docs.ansible.com/ansible/latest/dev_guide/developing_python_3.html#ansible-and-python-3) in the collection ``README.md`` and in every module and plugin (or in doc fragments)
- [x] only uses the [allowed plugin types](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst#modules-plugins) in the `plugins/` directory
- [x] has `README.md`
  **COMMENT** you could also put a link to a corresponding Matrix channel if exists and the badge like in [c.mysql](https://github.com/ansible-collections/community.mysql). Just a fyi
- [x] documentation, examples, and return sections use FQCNs for the `M(..)` [format macros](https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#linking-and-other-format-macros-within-module-documentation) when referring to modules
- [x] modules (or plugins) from ansible-core use `ansible.builtin.` as a FQCN prefix
- [x] documentation and return sections use `version_added:` containing the *collection* version for which an option, module or plugin was added
- [x] FQCNs are used for `extends_documentation_fragment:`, unless you are referring to doc_fragments from ansible-core
- [x] public plugins, roles and playbooks do not use files outside of `meta/`, `plugins/`, `roles/`, and `playbooks/`

**Collection management:**
- [x] `galaxy.yml` having `tags` field set
- [x] collection dependencies must have a lower bound on the version which is at least 1.0.0, and are all part of the `ansible` package
- [x] `meta/runtime.yml` defines the minimal version of Ansible which the collection works with
- [x] has changelog, preferably with `changelogs/changelog.yaml`
- [x] collection repository should not contain any large objects (binaries) comparatively to the current Galaxy tarball size limit of 20 MB. For example, package installers for testing purposes shouldn't be added.
- [x] collection repository should not contain any unnecessary files like, for example, temporary files. Temporary files should be added to `.gitignore`.
  **SHOULD FIX:** there are `tests/sanity/ignore*.txt` files that are empty, should be deleted i think. Also maybe `.vscode`? Anyway, this is `should fix` stuff, not must

**Tests:**

Note for reviewers: If you don't know how to check the points below, please ask maintainers of the collection you're reviewing how you can do it.
* In most cases, CI is set up via GitHub Actions.
* Check `.yml` files in the `.github/workflows` directory. There must be at least the `sanity` section under `jobs` containing the `ansible-test sanity` command running against all supported ansible-core versions that must be also listed there, for example, `- stable-2.11, - stable-2.12, - stable-2.13`.
* Check workflow runs by clicking the `Actions` tab in the repository's page - you're interested in `Scheduled` runs, runs against release commits and runs against pull requests.
* If there are no workflows in the `Actions` tab, ask the collection maintainers how CI is implemented.

- [x] passed `ansible-test sanity`
- [x] if `test/sanity/ignore*.txt` exists, it MUST not contain error codes listed [here](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst#ci-testing)
- [x] has CI tests up and running against each of the "major versions" of `ansible-base`/`ansible-core` that the collection supports
- [x] all CI tests MUST run against every pull request
- [x] all CI tests MUST run regularly (nightly, or at least once per week)
- [x] sanity tests MUST run against a commit that releases the collection; if they don't pass, the collection won't be released

**Reply by Andersson007** [reply]
@jborean93 i've finished the second review, could you please take a look?
please let me know when the feedback is incorporated or if you have any questions (I'll be on PTO next week, so will probably respond later).
Thanks

**Reply by jborean93** [reply]
>  has a policy of releasing, versioning and deprecation announced to contributors and users in some way
MUST FIX I see the process but don't see the policy. It can be in README or in an issue. See community.general as an example.

Can you share a link to this, `community.general` has this https://github.com/ansible/community-docs/blob/main/releasing_collections.rst but it just says moved to `docs.ansible.com`. Is simply linking to https://docs.ansible.com/ansible/devel/community/collection_contributors/collection_releasing.html enough? I'll extend the README to include references to semver though. Even then that seems to be a guide on how to approach collection versioning rather than something specific to `community.general`.

> MUST FIX please [reflect](https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#documentation-block) in the DOCUMENTATION that the check mode is supported by modules (where supported)

What modules in particular are missing this? All the modules in there have the following attributes.

```yaml
attributes:
  check_mode:
    support: full
```

> SHOULD FIX: there are tests/sanity/ignore*.txt files that are empty, should be deleted i think. Also maybe .vscode? Anyway, this is should fix stuff, not must

I don't see the harm in having the empty ignore files. Does it break anything you know? I like having them there because I know what versions of Ansible I've tested against and if I do end up having to ignore I don't need to create the files again.

I prefer not to remove the `.vscode` dir, it contains project settings specific to this collection like auto formatting rules for PowerShell and IDE PYTHONPATH settings.

**Reply by jborean93** [reply]
https://github.com/ansible-collections/microsoft.ad/pull/22 adds more info for the release policy, please let me know if you need any more info.

**Reply by Andersson007** [reply]
@jborean93 thanks for the response

1. Releasing info. Sorry, c.general has it in the [pinned issue](https://github.com/ansible-collections/community.general/issues/582). But what you added to readme, i think, is adequate. (though I like c.g's approach personally, so can be taken into consideration).

2. Check mode. TIL about attributes, thanks. Do you know any other included modules that use this? I'd like to see how it looks on docs.ansible.com and if it looks good I'll introduce it to the collections i maintain

3. ignore and vscode files. OK

**Reply by jborean93** [reply]
> Check mode. TIL about attributes, thanks. Do you know any other included modules that use this? I'd like to see how it looks on docs.ansible.com and if it looks good I'll introduce it to the collections i maintain

It's a relatively new thing and not very well documented. Here is an example of how it is rendered through the Ansible docs https://ansible-collections.github.io/microsoft.ad/branch/main/collections/microsoft/ad/computer_module.html#attributes

![image](https://user-images.githubusercontent.com/8462645/226557959-2c013ea3-4868-4498-86f6-d9617d28e576.png)

**Reply by Andersson007** [reply]
looks great, thank you for the example!

**Reply by briantist** [reply]
@Andersson007 I added attributes to the `community.hashi_vault` docs: https://docs.ansible.com/ansible/devel/collections/community/hashi_vault/vault_kv2_delete_module.html#attributes

I think most if not all of the collections @felixfontein maintains have them too :)

**Reply by Andersson007** [reply]
@briantist cool, thanks for the links! I filed add-attributes issues in the collections I maintain.

@thanks for taking a look at and approving the collection for inclusion!

**Comment by Andersson007** [comment]
+1 from me for inclusion

Could anyone from @ansible-collections/steering-committee approve the collection too?

**Reply by jborean93** [reply]
Thanks for the review, greatly appreciate it.

**Comment by briantist** [comment]
+1 for inclusion!

**Comment by mariolenz** [comment]
I've created ansible-community/community-topics#217.

If no one objects, we'll add the collection in a week: ansible-community/ansible-build-data#204

**Comment by russoz** [comment]
+1

**Comment by mariolenz** [comment]
> If no one objects, we'll add the collection in a week: ansible-community/ansible-build-data#204

Too late for ansible 7.4.0, but the collection has been added and should be part of 7.5.0.
